Terragrunt 101 (#6181)
gruntwork-io/terragrunt: docs/src/content/docs/02-guides/01-terragrunt-101/07-composing-stacks.md

## Passing Values to Units

Both units and stacks can receive **environment-specific configuration** via values.

The `values` attribute passes data into units:

```hcl
# terragrunt.stack.hcl
unit "vpc" {
  source = "../units/vpc"
  path   = "vpc"

  values = {
    name_prefix = "production"
    cidr_block  = "10.0.0.0/16"
  }
}
```

---

### Accessing Values in Units

Inside the unit's `terragrunt.hcl`, access these values via **`values.*`**:

```hcl
# units/vpc/terragrunt.hcl
terraform {
  source = "../../modules/vpc"
}

inputs = {
  name_prefix  = values.name_prefix
  vpc_settings = {
    cidr_block = values.cidr_block
  }
}
```

---

### Conditional Logic with Values

Values enable the same definition to **adapt to different contexts**:

```hcl
# units/vpc/terragrunt.hcl
locals {
  # Default to dev settings if not in a stack context
  name_prefix = try(values.name_prefix, "dev")
  cidr_block  = try(values.cidr_block, "10.0.0.0/16")
}

inputs = {
  name_prefix  = local.name_prefix
  vpc_settings = {
    cidr_block = local.cidr_block
  }
}
```

> > Using `try()` allows the unit to work **both** as a standalone deployment **and** as part of a stack.

Reviewer: ```suggestion
> Using `try()` allows the unit to define sensible defaults for values that don't _have_ to be defined explicitly for generation.
```

### Values Flow Through Stacks

The same pattern works at the stack level. A stack definition can accept values and pass them to its units:

```hcl
# stacks/vpc/terragrunt.stack.hcl

locals {
  # Accept values from parent, with defaults
  environment = try(values.environment, "dev")
  cidr_block  = try(values.cidr_block, "10.0.0.0/16")
}

unit "vpc" {
  source = "../../units/vpc"
  path   = "vpc"

  values = {
    name_prefix = local.environment
    cidr_block  = local.cidr_block
  }
}
```

This creates a **single definition** that can be instantiated with different values for each environment.

## Working with Stack Commands

Explicit stacks introduce new commands under **`terragrunt stack`**:

| Command                            | Description                              |
| :--------------------------------- | :--------------------------------------- |
| **`terragrunt stack generate`**    | Generate units from the stack definition |
| **`terragrunt stack run plan`**    | Plan all units in the stack              |
| **`terragrunt stack run apply`**   | Apply all units in the stack             |
| **`terragrunt stack run destroy`** | Destroy all units in reverse order       |

> | **`terragrunt stack run output`**  | Show outputs from all units              |

Reviewer: ```suggestion
| **`terragrunt stack output`**  | Show consolidated outputs from all units              |
```

---

### The Generate Step

> Before operating on an explicit stack, you must **generate it**:

Reviewer: ```suggestion
Before operating on an explicit stack, you can explicitly **generate it**:

Stacks are generated automatically by default for all commands that interact with stacks.
```

That second sentence can be a Tip Aside, but that requires renaming the file and adding the import.

```bash
cd my-stack/
terragrunt stack generate
```

This creates the directory structure defined in your `terragrunt.stack.hcl`. Each unit gets its own directory with a `terragrunt.hcl` file.

> When a stack includes a `stack` block (referencing another stack), the referenced stack's units are also generated.

Reviewer: ```suggestion
When a stack defines a `stack` block (referencing another stack), the referenced stack's units are also generated.
```

---

### Generated File Management

> By default, `terragrunt stack generate` creates generated units in their `path` directories. A common pattern is to:

Reviewer: ```suggestion
By default, `terragrunt stack generate` creates generated units in their `path` directories nested under a `.terragrunt-stack` directory.

A common pattern is to:
```

1. Keep stack definitions in version control

> 2. Add generated files to `.gitignore`

Reviewer: ```suggestion
2. Add `.terragrunt-stack` to `.gitignore`
```

## Nested Stacks

Stacks can include **other stacks**, enabling composition at a higher level:

```hcl
# terragrunt.stack.hcl

unit "vpc" {
  source = "../units/vpc"
  path   = "vpc"
}

stack "monitoring" {
  source = "../stacks/monitoring"
  path   = "monitoring"
}

stack "logging" {
  source = "../stacks/logging"
  path   = "logging"
}
```

> The `stack` block works similarly to `unit`, but references another `terragrunt.stack.hcl` file instead of a `terragrunt.hcl` file.

---

### Stack Block Attributes

These are the most important attributes to remember when working with [`stack`](https://docs.terragrunt.com/reference/hcl/blocks/#stack) blocks.

| Attribute    | Required | Description                               |
| :----------- | :------- | :---------------------------------------- |
| **`source`** | Yes      | Path to another stack definition          |
| **`path`**   | Yes      | Local directory path for the nested stack |
| **`values`** | No       | Values to pass to the nested stack        |

---

### Building a Stack Catalog

Nested stacks enable building **reusable infrastructure blueprints**:

```text

> catalog/

Reviewer: This should be a FileTree

├── stacks/
│   ├── eks-cluster/
│   │   └── terragrunt.stack.hcl
│   ├── monitoring/
│   │   └── terragrunt.stack.hcl
│   └── full-environment/
│       └── terragrunt.stack.hcl  # Composes eks-cluster + monitoring
└── units/
    ├── vpc/
    ├── eks/
    └── prometheus/
```

| Level                | Purpose                              |
| :------------------- | :----------------------------------- |
| **Units**            | Individual infrastructure components |
| **Stacks**           | Composed collections of units        |
| **Full environment** | Composed collections of stacks       |

## Environment Pattern with Explicit Stacks

A powerful pattern is defining a **single stack blueprint** and deploying it **multiple times** with different configurations:

```text

> project/

Reviewer: This should be a FileTree

  path   = "vpc"
  values = {
    environment = local.environment
    vpc_cidr    = local.vpc_cidr
  }
}

unit "vpc_subnet" {
  source = "${local.units_dir}/vpc-subnet"
  path   = "vpc-subnet"
  values = {
    environment = local.environment
    region      = local.region
    azs         = local.azs
  }
}
# ... more units (routes, etc.)
```

---

### Stack Instantiation Files

Each instantiation creates a small file that references the blueprint and passes values:

#### First VPC (vpc-dev1)

```hcl
# infrastructure-live/dev/us-east-1/vpc-dev1/terragrunt.stack.hcl
stack "network" {
  source = "${get_repo_root()}/../stacks/three-tier-vpc"
  path   = "."
  values = {
    environment = "dev1"
    vpc_cidr    = "10.0.0.0/16"
    region      = "us-east-1"
    azs         = ["us-east-1a", "us-east-1b"]
  }
}
```

#### Second VPC (vpc-dev2)

```hcl
# infrastructure-live/dev/us-east-1/vpc-dev2/terragrunt.stack.hcl
stack "network" {
  source = "${get_repo_root()}/../stacks/three-tier-vpc"
  path   = "."
  values = {
    environment = "dev2"
    vpc_cidr    = "10.1.0.0/16"
    region      = "us-east-1"
    azs         = ["us-east-1a", "us-east-1b"]
  }
}
```

> > These files are simple. They just specify the `source` and instance-specific `values`.

Reviewer: This should be an Aside.
